### DISCUSSION TITLE
FBL / ARF support

### DISCUSSION THREAD

**Discussion by alexroosenstein** [postalserver/postal#1515] [Feature suggestions]
Would be great if Postal could support feedback loop reports.

Perhaps even with a webhook event when a specific email is marked as spam.

**Comment by adamcooke** [comment]
I agree. This would be useful. I've marked this as an enhancement. Hopefully someone will be able to look at this in the near future.

**Comment by rsgehlot** [comment]
Do we any update on this, these days fbl is kind of very critical and mandatory to prevent your system..

**Comment by navinp** [comment]
Is there anyone here that can do this, even if I have to hire?

**Comment by gazugafan** [comment]
I dug into this a bit today. Unfortunately, from what I can tell, the Validity ARF reports don't include the original message, headers, or even the email address of who sent the spam report. So there's no way to link the spam report back to the offending message.

What they do include is the original Return-Path. I'm guessing this is what Amazon SES, SendGrid, etc work off of. Those services use a unique message ID as the username in the Return-Path, and use that for linking bounces (and spam reports, it seems) back to the original message.

Postal identifies bounces using a special X-Postal-MsgID header that is (usually) repeated in the bounce. But again, none of the original headers besides the Return-Path are repeated in the Validity ARF reports.

So, the first step in making this work would be to switch to using a unique message ID in the Return-Path username instead of (or in addition to) the X-Postal-MsgID header.

**Reply by willpower232** [reply]
I was under the impression it was already unique but I can't figure out where it is in the database, I've seen that Postal is aware of a message having a bounce so I believe there must be something but I'm not sure how it works

**Reply by gazugafan** [reply]
If you look at the Return-Path username in an email from Postal, you might mistakenly think that it's unique (I certainly did at first). It seems to be a random string, but it's actually only unique to the _**server**_... not the message. So all the emails from the same "server" will have the same Return-Path. This is the "server" as configured in Postal. It seems to be the ```token``` field, which you'll find in the ```servers``` table. No idea if it's truly just a random string, or if it's something encoded.

To identify bounces, it seems Postal includes an ```X-Postal-MsgID``` header in each email containing the actual message ID. This is the ```token``` field in the ```messages``` table of the server's database. The problem is that not all bounces will include the headers from the original email, so sometimes a bounce will be received without the ```X-Postal-MsgID``` header to identify which message bounced. And, of course I found that Validity ARF reports don't include the original headers either (just the Return-Path), so it's the same problem trying to identify which message a spam report is for.

**Reply by willpower232** [reply]
ah that makes sense since the messages are stored by server so the resolution would be a two-part username to include server and message ID relatively immutably.

I've definitely seen bounces in Postal where it couldn't find the original message which makes more sense now.
